cpp/src/stream_compaction/distinct_helpers.cu

```diff
@@ -16,96 +16,125 @@
 
-#include <cudf/detail/hash_reduce_by_row.cuh>
-
-#include <rmm/resource_ref.hpp>
-
 namespace cudf::detail {
 
-namespace {
-/**
- * @brief The functor to find the first/last/all duplicate row for rows that compared equal.
- */
-template <typename MapView, typename KeyHasher, typename KeyEqual>
-struct reduce_fn : reduce_by_row_fn_base<MapView, KeyHasher, KeyEqual, size_type> {
-  duplicate_keep_option const keep;
-
-  reduce_fn(MapView const& d_map,
-            KeyHasher const& d_hasher,
-            KeyEqual const& d_equal,
-            duplicate_keep_option const keep,
-            size_type* const d_output)
-    : reduce_by_row_fn_base<MapView, KeyHasher, KeyEqual, size_type>{d_map,
-                                                                     d_hasher,
-                                                                     d_equal,
-                                                                     d_output},
-      keep{keep}
-  {
+template <typename RowHasher>
+rmm::device_uvector<size_type> process_keep_option(hash_set_type<RowHasher>& set,
+                                                   size_type set_size,
+                                                   size_type num_rows,
+                                                   duplicate_keep_option keep,
+                                                   rmm::cuda_stream_view stream,
+                                                   rmm::device_async_resource_ref mr)
+{
+  auto output_indices = rmm::device_uvector<size_type>(set_size, stream, mr);
+
+  // If we don't care about order, just gather indices of distinct keys taken from set.
+  if (keep == duplicate_keep_option::KEEP_ANY) {
+    set.retrieve_all(output_indices.begin(), stream.value());
+    return output_indices;
   }
 
-  __device__ void operator()(size_type const idx) const
-  {
-    auto const out_ptr = this->get_output_ptr(idx);
-
-    if (keep == duplicate_keep_option::KEEP_FIRST) {
-      // Store the smallest index of all rows that are equal.
-      atomicMin(out_ptr, idx);
-    } else if (keep == duplicate_keep_option::KEEP_LAST) {
-      // Store the greatest index of all rows that are equal.
-      atomicMax(out_ptr, idx);
-    } else {
-      // Count the number of rows in each group of rows that are compared equal.
-      atomicAdd(out_ptr, size_type{1});
+  auto reduction_results = rmm::device_uvector<size_type>(num_rows, stream, mr);
+  thrust::uninitialized_fill(rmm::exec_policy(stream),
+                             reduction_results.begin(),
+                             reduction_results.end(),
+                             reduction_init_value(keep));
+
+  auto set_ref = set.ref(cuco::op::insert_and_find);
+
+  thrust::for_each(rmm::exec_policy(stream),
+                   thrust::make_counting_iterator(0),
+                   thrust::make_counting_iterator(num_rows),
+                   [set_ref, keep, reduction_results = reduction_results.begin()] __device__(
```

Comment:
Wrap device lambda in `cuda::proclaim_return_type`.

Reply:
Out of curiosity, @ttnghia is the `proclaim_return_type` wrapper really needed in this case? The code seems to build just fine without it.

Reply:
In the past it was recommended to add that qualifier to all device lambda in thrust calls. Since CCCL is not fully developed, we don't know if this can compile now but won't compile in the future thus we should better be conservative.

Reply:
It makes sense if the function returns something.

Given that the return type is `void`, I'm inclined to retain it as is.